### PR TITLE
DEV: updated KM hashing algorithm -> Nested serialization

### PR DESCRIPTION
**Pull request by lacabra**
Deprecating #150 in favor of this PR that has a branch name matching enigmampc/enigma-core#217, but otherwise has the same edits.

### CODE REVIEW

File: contracts/impl/EnigmaCommon.sol [hunk 1]
```diff
@@ -101,4 +119,20 @@ library EnigmaCommon {
...
+    * @param _arraylength Length of array
+    * @param _typeLength Length of each item in bytes
+    * @return New bytes buffer
+    */
+    function appendMessageArrayLengthKM(uint256 _arraylength, uint256 _typeLength, bytes memory _message)
+    internal
+    pure
+    returns (bytes memory)
+    {
+        return (_message.concat(hex"01")).concat(uint64(_arraylength.mul(_typeLength.add(9))).toBytesFromUint64());
```
Comment (moriaab) [inline contracts/impl/EnigmaCommon.sol]: I think that the hardcoded solution is never good. It should be a size of byte and size of `uint64`.

File: contracts/impl/EnigmaCommon.sol [hunk 2]
```diff
@@ -87,6 +90,21 @@ library EnigmaCommon {
...
+    * @param _message Bytes buffer being appended to
+    * @param _var Bytes representation of value that needs to be concatenated to existing buffer
+    * @return New bytes buffer
+    */
+    function appendMessageKM(bytes memory _message, bytes memory _var)
+    internal
+    pure
+    returns (bytes memory)
+    {
+        return ((_message.concat(hex"00")).concat(uint64(_var.length).toBytesFromUint64())).concat(_var);
```
Comment (AvishaiW) [inline contracts/impl/EnigmaCommon.sol:105]: does the length include the 00 byte as well?

Comment (apalepu23) [reply contracts/impl/EnigmaCommon.sol:105]: no, does not, but i believe the solution on core end also works same way as this now. should be resolved!

Comment (lacabra) [reply contracts/impl/EnigmaCommon.sol:105]: @AvishaiW: this appears to be the matching code in `core`:
https://github.com/enigmampc/enigma-core/blob/77f48bc744c7d15a30fb42940c9a7d4987dcb351/enigma-principal/enclave/src/epoch_keeper_t/nested_encoding.rs#L20

As we can see the `0` byte is not taken into account in computing the length. Is this the intended behavior? As per what is documented in https://github.com/enigmampc/protocol-discovery/blob/master/docs/hash_mul_nested.pdf at the beginning describing `Serialization Function`, that seems indeed to be the intended behavior.

Still, I want confirmation from @moriaab before merging this.

**Review by apalepu23 — APPROVED** [review]

**Pull request merged**